lib/app/migrate.ts

```
@@ -0,0 +1,584 @@
+import { UNMIGRATABLE_REASONS } from '@hubspot/local-dev-lib/constants/projects';
+import { mapToUserFacingType } from '@hubspot/project-parsing-lib/src/lib/transform';
+
+function getUnmigratableReason(reasonCode: string) {
+  switch (reasonCode) {
+    case UNMIGRATABLE_REASONS.UP_TO_DATE:
+      return i18n(
+        'commands.project.subcommands.migrateApp.migrationNotAllowedReasons.upToDate'
+      );
+    case UNMIGRATABLE_REASONS.IS_A_PRIVATE_APP:
+      return i18n(
+        'commands.project.subcommands.migrateApp.migrationNotAllowedReasons.isPrivateApp'
+      );
+    case UNMIGRATABLE_REASONS.LISTED_IN_MARKETPLACE:
+      return i18n(
+        'commands.project.subcommands.migrateApp.migrationNotAllowedReasons.listedInMarketplace'
+      );
+    default:
+      return i18n(
+        'commands.project.subcommands.migrateApp.migrationNotAllowedReasons.generic',
+        {
+          reasonCode,
+        }
+      );
+  }
+}
+
+async function handleMigrationSetup(
+  derivedAccountId: number,
+  options: ArgumentsCamelCase<MigrateAppOptions>
+) {
+  const { name, dest, appId } = options;
+  const { data } = await listAppsForMigration(derivedAccountId);
+
+  const { migratableApps, unmigratableApps } = data;
+  const allApps = [...migratableApps, ...unmigratableApps];
+
+  if (allApps.length === 0) {
+    throw new Error(
+      i18n(`commands.project.subcommands.migrateApp.errors.noApps`, {
+        accountId: derivedAccountId,
+      })
+    );
+  }
+
+  if (migratableApps.length === 0) {
+    const reasons = unmigratableApps.map(
+      app =>
+        `${chalk.bold(app.appName)}: ${getUnmigratableReason(app.unmigratableReason)}`
+    );
+
+    throw new Error(
+      `${i18n(`commands.project.subcommands.migrateApp.errors.noAppsEligible`, {
+        accountId: derivedAccountId,
+      })} \n  - ${reasons.join('\n  - ')}`
+    );
+  }
+
+  if (
+    appId &&
+    !allApps.some(app => {
+      return app.appId === appId;
+    })
+  ) {
+    throw new Error(
```

Do we want to throw an error here, before the user chooses the app to migrate? Would that be confusing?

---

This is checking if the user provided and appId that is not in the list of apps, so they have already chosen an app.